src/helpers/check_diff.jl

```diff
@@ -0,0 +1,136 @@
+            df[counter, :r2] = r2
+            df[counter, :left] = x[i]
+            df[counter, :right] = x[j]
+            df[counter, :length] = x[j] - x[i]
+            df[counter, :left_idx] = i
+            df[counter, :right_idx] = j
+            counter += 1
+        end
+    end
+
+    df = df[df[:, :r2] .> r2_cutoff, :]
+    df = sort(df, :length; rev=true)
+
+    res = DataFrame(;
+        a=zeros(seq_len),
+        b=zeros(seq_len),
+        r2=zeros(seq_len),
+        left=zeros(seq_len),
+        right=zeros(seq_len),
+        length=zeros(seq_len),
+        left_idx=zeros(Int32, seq_len),
+        right_idx=zeros(Int32, seq_len),
+    )
+    counter = 1
+    covered_set = collect(1:seq_len)
+    while size(covered_set, 1) > 0
+        line = df[1, :]
+        res[counter, :] = line
+        counter += 1
+        remove_rows =
+            (line.left .<= df[:, :left] .< line.right) .|
+            (line.left .< df[:, :right] .<= line.right)
+        remove_rows[1] = true
+        df = df[.!remove_rows, :]
+        setdiff!(covered_set, collect((line.left_idx):(line.right_idx)))
+    end
+    res = res[res[:, :left_idx] .!= 0, :]
+    res = sort(res, :left)
+
+    return res
+end
+
+function check_diff(
+    M::AbstractManifold,
+    cost_fnc,
+    diff_fnc;
+    p=random_point(M),
+    X=random_tangent(M, p),
+    log_space_range=range(-8, 0; length=101),
+    plot=false,
+)
+    check_point(M, p)
+    check_vector(M, p, X)
+    f0 = cost_fnc(M, p)
+    df0 = diff_fnc(M, p, X)
+
+    log_space = 10 .^ log_space_range
+
+    # Based on the Taylor approximation of development of the cost function around point x:
+    # f(R_x(t * X)) = f(x) + t * <grad f(x), X)> + O(t^2)
+    lhs = [cost_fnc(M, geodesic(M, p, X, t)) for t in log_space]
+    rhs = [f0 + t * df0 for t in log_space]
+    err = abs.(rhs - lhs)
+
+    df = linearize_function(log10.(log_space), log10.(err))
+    total_quadratic_length = sum(df[1.8 .< df[:, :b], :length])
+    mostly_quadratic =
+        total_quadratic_length >
+        0.75 * (maximum(log_space_range) - minimum(log_space_range))
+    if plot
+        plt = Plots.scatter(
+            log10.(log_space),
+            log10.(err);
+            legend=false,
+            title=if mostly_quadratic
+                "The differential seems correct"
+            else
+                "The differential seems incorrect"
+            end,
+            xlabel="log10 distance to p",
+            ylabel="log10 approximation error of differential",
+        )
+        Plots.abline!(plt, 1, 0; color="red")
+        Plots.abline!(plt, 2, 0; color="green")
+        for i in 1:size(df, 1)
+            xs = [df[i, :left], df[i, :right]]
+            Plots.plot!(
+                plt,
+                xs,
+                df[i, :a] .+ xs .* df[i, :b];
+                color=df[i, :b] > 1.8 ? "green" : "red",
+            )
+        end
+        return plt
+    else
+        if !mostly_quadratic
```

Comment:
returning an exception is not that good here, we do that for our internal checks (but they are called by `is:_point`/`is_vector` which throw them.
We could print a warning or
we could do a similar p´boolean for prints or even to an `kwarg` of `print::Union{IO,Nothing}=nothing` (which one can set to `stdout`  to get a print detail.